File: acados/cartPole/model_cartpole.py
"""
model_cartpole.py
Modelo del péndulo invertido sobre carrito para Acados
"""

from acados_template import AcadosModel
import casadi as ca

def model_cartpole():
    """
    Modelo del péndulo invertido sobre carrito
    Estados: x = [x, theta, x_dot, theta_dot]
    Control: u = [F]
    Parámetros: p = [m_c, m_p, l, I_p, g]
    """
    
    # Crear objeto AcadosModel (NO un diccionario)
    model = AcadosModel()
    model.name = "cartpole_mujoco"
    
    # =========================
    # Parámetros simbólicos
    # =========================
    m_c = ca.SX.sym("m_c")   # masa carro [kg]
    m_p = ca.SX.sym("m_p")   # masa péndulo [kg]
    l   = ca.SX.sym("l")     # longitud al COM péndulo [m]
    I_p = ca.SX.sym("I_p")   # inercia péndulo (eje hinge) [kg⋅m²]
    g   = ca.SX.sym("g")     # gravedad [m/s²]

    p = ca.vertcat(m_c, m_p, l, I_p, g)

    # =========================
    # Estados
    # =========================
    x      = ca.SX.sym("x")          # posición del carro [m]
    theta  = ca.SX.sym("theta")      # ángulo del péndulo [rad]
    x_dot  = ca.SX.sym("x_dot")      # velocidad del carro [m/s]
    th_dot = ca.SX.sym("theta_dot")  # velocidad angular [rad/s]

    states = ca.vertcat(x, theta, x_dot, th_dot)

    # =========================
    # Control
    # =========================
    F = ca.SX.sym("F")  # fuerza aplicada [N]
    controls = ca.vertcat(F)

    # =========================
    # Dinámica
    # =========================
    sin_t = ca.sin(theta)
    cos_t = ca.cos(theta)

    # Denominador común
    denom = m_c + m_p - (m_p**2 * l**2 * cos_t**2) / (I_p + m_p * l**2)

    # Aceleración del carro
    x_ddot = (
        F
        + m_p * l * sin_t * th_dot**2
        - m_p * g * sin_t * cos_t
    ) / denom

    # Aceleración angular del péndulo
    theta_ddot = (
        m_p * g * l * sin_t
        - m_p * l * cos_t * x_ddot
    ) / (I_p + m_p * l**2)

    # Ecuaciones de estado explícitas: dx/dt = f(x, u, p)
    f_expl = ca.vertcat(
        x_dot,
        th_dot,
        x_ddot,
        theta_ddot
    )

    # =========================
    # Forma implícita (requerida por algunos integradores)
    # =========================
    # Variables para las derivadas temporales
    x_dot_sym      = ca.SX.sym("x_dot_sym")
    theta_dot_sym  = ca.SX.sym("theta_dot_sym")
    x_ddot_sym     = ca.SX.sym("x_ddot_sym")
    theta_ddot_sym = ca.SX.sym("theta_ddot_sym")
    
    xdot = ca.vertcat(x_dot_sym, theta_dot_sym, x_ddot_sym, theta_ddot_sym)
    
    # Forma implícita: 0 = xdot - f(x, u, p)
    f_impl = xdot - f_expl

    # =========================
    # Asignar al modelo de Acados
    # =========================
    model.x = states
    model.xdot = xdot
    model.u = controls
    model.p = p
    model.f_expl_expr = f_expl
    model.f_impl_expr = f_impl

    return model  # Retornar el objeto AcadosModel, NO un diccionario


# Test del modelo
if __name__ == "__main__":
    print("="*60)
    print("Probando creación del modelo del cartpole")
    print("="*60)
    
    model = model_cartpole()
    
    print(f"✅ Modelo creado: {model.name}")
    print(f"   Estados (nx):     {model.x.size()[0]}")
    print(f"   Controles (nu):   {model.u.size()[0]}")
    print(f"   Parámetros (np):  {model.p.size()[0]}")
    print(f"\n✅ El modelo está listo para Acados!")
    print("="*60)

File: Projects/Matlab/getting_started/model_cart_pole.m
function model = model_cart_pole(varargin)
    import casadi.*
    
    %% Parámetros del sistema
    M = 7.67;        % masa del carrito [kg]
    m = 0.7;         % masa del péndulo [kg]
    l = 0.06;        % longitud al CDG del péndulo [m]
    g = 9.81;        % gravedad [m/s^2]
    
    % Momento de inercia del péndulo
    I_pend_gmm2 = 3.975E+06;  % Iyy en g·mm²
    I_pend = I_pend_gmm2 * 1E-9;  % Conversión a kg·m²
    
    % Momento de inercia total sobre el pivote
    J = m*l^2 + I_pend;
    
    %% Variables simbólicas
    p = SX.sym('p');
    theta = SX.sym('theta');
    v = SX.sym('v');
    dtheta = SX.sym('dtheta');
    F = SX.sym('F');
    
    x = vertcat(p, theta, v, dtheta);
    xdot = SX.sym('xdot', 4, 1);
    u = F;
    
    %% Dinámica CORREGIDA
    sin_theta = sin(theta);
    cos_theta = cos(theta);
    
    denom = (M + m)*J - (m*l*cos_theta)^2;
    
    % CORRECCIÓN 1: Signo del término gravitacional en p_ddot
    p_ddot = (J*F + m*l*sin_theta*(J*dtheta^2 + m*g*l*cos_theta)) / denom;
    
    % CORRECCIÓN 2: Signo del término centrífugo en theta_ddot
    theta_ddot = (-m*l*cos_theta*F - (M+m)*m*g*l*sin_theta ...
                  - (m*l)^2*sin_theta*cos_theta*dtheta^2) / denom;
    
    f_expl_expr = vertcat(v, dtheta, p_ddot, theta_ddot);
    f_impl_expr = f_expl_expr - xdot;
    
    %% Modelo
    model = AcadosModel();
    model.x = x;
    model.xdot = xdot;
    model.u = u;
    model.f_expl_expr = f_expl_expr;
    model.f_impl_expr = f_impl_expr;
    model.name = 'cart_pole';
end